src/pp/if_parser.rs

```diff
@@ -0,0 +1,441 @@
+
+                match punct {
+                    Punct::LeftAngle => left = (left < right) as i64,
+                    Punct::RightAngle => left = (left > right) as i64,
+                    Punct::LessEqual => left = (left <= right) as i64,
+                    Punct::GreaterEqual => left = (left >= right) as i64,
+                    _ => unreachable!(),
+                }
+            } else {
+                break;
+            }
+        }
+
+        Ok(left)
+    }
+
+    fn parse_equality(&mut self) -> Step<i64> {
+        let mut left = self.parse_comparative()?;
+
+        while let Some(TokenValue::Punct(punct)) = self.peek()?.map(|t| t.value) {
+            if let Punct::Equal | Punct::NotEqual = punct {
+                self.next()?;
+
+                let right = self.parse_comparative()?;
+
+                match punct {
+                    Punct::Equal => left = (left == right) as i64,
+                    Punct::NotEqual => left = (left != right) as i64,
+                    _ => unreachable!(),
+                }
+            } else {
+                break;
+            }
+        }
+
+        Ok(left)
+    }
+
+    fn parse_bit_and(&mut self) -> Step<i64> {
+        let mut left = self.parse_equality()?;
+
+        while let Some(TokenValue::Punct(Punct::Ampersand)) = self.peek()?.map(|t| t.value) {
+            self.next()?;
+
+            let right = self.parse_equality()?;
+
+            left &= right
+        }
+
+        Ok(left)
+    }
+
+    fn parse_bit_xor(&mut self) -> Step<i64> {
+        let mut left = self.parse_bit_and()?;
+
+        while let Some(TokenValue::Punct(Punct::Caret)) = self.peek()?.map(|t| t.value) {
+            self.next()?;
+
+            let right = self.parse_bit_and()?;
+
+            left ^= right
+        }
+
+        Ok(left)
+    }
+
+    fn parse_bit_or(&mut self) -> Step<i64> {
+        let mut left = self.parse_bit_xor()?;
+
+        while let Some(TokenValue::Punct(Punct::Pipe)) = self.peek()?.map(|t| t.value) {
+            self.next()?;
+
+            let right = self.parse_bit_xor()?;
+
+            left |= right
+        }
+
+        Ok(left)
+    }
+
+    fn parse_logical_and(&mut self) -> Step<i64> {
+        let mut left = self.parse_bit_or()?;
+
+        while let Some(TokenValue::Punct(Punct::LogicalAnd)) = self.peek()?.map(|t| t.value) {
+            self.next()?;
+
+            let right = self.parse_bit_or()?;
+
+            left = (left != 0 && right != 0) as i64;
+        }
+
+        Ok(left)
+    }
+
+    fn parse_logical_or(&mut self) -> Step<i64> {
+        let mut left = self.parse_logical_and()?;
+
+        while let Some(TokenValue::Punct(Punct::LogicalAnd)) = self.peek()?.map(|t| t.value) {
+            self.next()?;
+
+            let right = self.parse_logical_and()?;
+
+            left = (left != 0 || right != 0) as i64;
+        }
+
+        Ok(left)
+    }
+
+    pub fn evaluate(&mut self) -> Step<bool> {
```

Comment:
The only thing we can do with an if_parser is call this function, so maybe it should just be a freestanding function and create an object internally.

In the future we'll also need to evaluate expressions for `#line` statements so this could be `evaluate_expression` that returns an `i64`.